cli/repair/clusterwide_config.go

```diff
@@ -0,0 +1,302 @@
+}
+
+// INSTANCES
+
+func getInstancesConf(topologyConf *TopologyConfType) (*map[string]InstanceConfType, error) {
+	instancesConfRaw, found := topologyConf.Raw[keyInstances]
+	if !found {
+		return nil, fmt.Errorf("Topology config doesn't contain %q key", keyInstances)
+	}
+
+	instancesConfRawMap, ok := instancesConfRaw.(RawConfType)
+	if !ok {
+		return nil, fmt.Errorf("%q value isn't a map", keyInstances)
+	}
+
+	topologyConf.InstancesRaw = instancesConfRawMap
+
+	instancesConf := make(map[string]InstanceConfType)
+
+	for instanceUUIDRaw, instanceConfRaw := range instancesConfRawMap {
+		instanceUUID, ok := instanceUUIDRaw.(string)
+		if !ok {
+			return nil, fmt.Errorf("Instance UUID isn't a string")
+		}
+
+		var instanceConf InstanceConfType
+
+		switch conf := instanceConfRaw.(type) {
+		case string:
+			if conf != expelledState {
+				return nil, fmt.Errorf("Instance %s is in the unknown state %s", instanceUUID, conf)
+			}
+			instanceConf.IsExpelled = true
+		case RawConfType:
+			isDisabled, ok := conf[keyInstanceDisabled]
+			if !ok {
+				return nil, fmt.Errorf("Instance %s config doesn't contain %q key", instanceUUID, keyInstanceDisabled)
+			}
+			instanceConf.IsDisabled, ok = isDisabled.(bool)
+			if !ok {
+				return nil, fmt.Errorf("Instance %s has %q that isn't a bool", instanceUUID, keyInstanceDisabled)
+			}
+
+			advertiseURI, ok := conf[keyInstanceAdvertiseURI]
+			if !ok {
+				return nil, fmt.Errorf("Instance %s config doesn't contain %q key", instanceUUID, keyInstanceAdvertiseURI)
+			}
+			instanceConf.AdvertiseURI, ok = advertiseURI.(string)
+			if !ok {
+				return nil, fmt.Errorf("Instance %s has %q that isn't a string", instanceUUID, keyInstanceAdvertiseURI)
+			}
+
+			replicasetUUID, ok := conf[keyInstanceReplicasetUUID]
+			if !ok {
+				return nil, fmt.Errorf("Instance %s config doesn't contain %q key", instanceUUID, keyInstanceReplicasetUUID)
+			}
+			instanceConf.ReplicasetUUID, ok = replicasetUUID.(string)
+			if !ok {
+				return nil, fmt.Errorf("Instance %s has %q that isn't a string", instanceUUID, keyInstanceReplicasetUUID)
+			}
+
+			instanceConf.Raw = conf
+		default:
+			return nil, fmt.Errorf("Instance %s config isn't a map or a string", instanceUUID)
+		}
+
+		instancesConf[instanceUUID] = instanceConf
+	}
+
+	return &instancesConf, nil
+}
+
+func setInstanceURIRaw(topologyConf *TopologyConfType, instanceUUID, newURI string) error {
+	instanceConf, ok := topologyConf.Instances[instanceUUID]
+	if !ok {
+		return fmt.Errorf("Instance %s isn't found in cluster", instanceUUID)
+	}
+
+	if instanceConf.IsExpelled {
+		return fmt.Errorf("Instance %s is expelled", instanceUUID)
+	}
+
+	instanceConf.Raw[keyInstanceAdvertiseURI] = newURI
+
+	return nil
+}
+
+func removeInstanceFromRaw(topologyConf *TopologyConfType, instanceUUID string) {
+	delete(topologyConf.InstancesRaw, instanceUUID)
+}
+
+// REPLICASETS
+
+func getReplicasetsConf(topologyConf *TopologyConfType) (*map[string]ReplicasetConfType, error) {
```

Comment:
It's too hard to understand code then validation logic is mixed with business logic. Please split them for all cases.

Reply:
You mean field types validation? I don't see any reasons to separate them from type casts.

`clusterwide_config` functions hide raw config details from business logic, and I sure it's enough.